File: 00948. Bag of Tokens/00948.go
// https://leetcode.com/problems/bag-of-tokens

func bagOfTokensScore(tokens []int, power int) int {
    sort.Ints(tokens)

    score, l, r := 0, 0, len(tokens)-1
    for l <= r {
        if power >= tokens[l] { // can play Face-up
            score++
            power -= tokens[l]
            l++
        } else if l < r && score > 0 {  // can play Face-down, and has opportunity to get more scores
            score--
            power += tokens[r]
            r--
        } else {    // end
            break
        }
    }

    return score
}

File: LeetCode/00948. Bag of Tokens/00948.c
int compare(const void *a, const void *b) {
    return *(int *)a - *(int *)b;
}

int bagOfTokensScore(int* tokens, int tokensSize, int power) {
    if (tokensSize == 0) {
        return 0;
    }

    qsort(tokens, tokensSize, sizeof(int), compare);

    int max = 0, score = 0, l = 0, r = tokensSize-1;

    while (l <= r) {
        if (power >= tokens[l]) {
            power -= tokens[l++];
            ++score;
            max = (max > score) ? max : score;
        } else if (score > 0) {
            power += tokens[r--];
            --score;
        } else {
            break;
        }
    }

    return max;
}
